compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,224 @@
+use crate::transform::MirPass;
+use crate::util::patch::MirPatch;
+use rustc_data_structures::stable_map::FxHashMap;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Const, List, Ty, TyCtxt};
+use rustc_span::def_id::DefId;
+use rustc_target::abi::{Size, TagEncoding, Variants};
+
+/// A pass that seeks to optimize unnecessary moves of large enum types, if there is a large
+/// enough discrepanc between them
+pub struct EnumSizeOpt<const DISCREPANCY: u64>;
+
+impl<'tcx, const D: u64> MirPass<'tcx> for EnumSizeOpt<D> {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        self.optim(tcx, body);
+    }
+}
+
+impl<const D: u64> EnumSizeOpt<D> {
+    fn candidate<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        ty: Ty<'tcx>,
+        body_did: DefId,
+    ) -> Option<(Size, u64, Vec<Size>)> {
+        match ty.kind() {
+            ty::Adt(adt_def, _substs) if adt_def.is_enum() => {
+                let p_e = tcx.param_env(body_did);
+                // FIXME(jknodt) handle error better below
+                let layout = tcx.layout_of(p_e.and(ty)).unwrap();
+                let variants = &layout.variants;
+                match variants {
+                    Variants::Single { .. } => None,
+                    Variants::Multiple { variants, .. } if variants.len() <= 1 => None,
+                    Variants::Multiple { tag_encoding, .. }
+                        if matches!(tag_encoding, TagEncoding::Niche { .. }) =>
+                    {
+                        None
+                    }
+                    Variants::Multiple { variants, .. } => {
+                        let min = variants.iter().map(|v| v.size).min().unwrap();
+                        let max = variants.iter().map(|v| v.size).max().unwrap();
+                        if max.bytes() - min.bytes() < D {
```

Comment:
The `max - min` criterion is based on the most optimistic scenario.

Consider two enums with the same maximum absolute deviation in variant sizes. The first enum has one large outlier and remaining variants are small, the second enum conversely has one small outlier and remaining variants are large. Both enums would be be equally good candidates under current criterion, but the first one would seem like a much better candidate.

What about assuming that variants are uniformly distributed and calculating expected reduction in the number of bytes copied?

Reply:
oops I never directly addressed this, I'm not sure if uniform distribution makes a ton of sense, as I'd expect if this is hit that one variant is probably significantly larger than the others. Probably something to experiment with.